### DISCUSSION TITLE
Delete shaping points without calculation

### DISCUSSION THREAD

**Discussion by 0709wiwiwi** [devemux86/cruiser#543] [Q&A] [answered]
Dealing with an excessive number of **_shaping trackpoints_.** 
It is more convenient to be able to edit a less crowded navtrack.
Action to do: 
Convert existing **_shaping trackpoints_** without recalculation  to simple trackpoints.
 
Reference: https://github.com/devemux86/cruiser/discussions/533#discussioncomment-9477157

**Comment by devemux86** [comment]
@0709wiwiwi thanks for the suggestion.

Routes need shaping points to closely follow the original tracks.
Shaping points can auto hide in the routing appearance settings.

**Comment by devemux86** [comment]
@0709wiwiwi thanks for the extra details.

So delete single shaping point will be done without route calculation.
We can use the recalculate section or route to do a new calculation.

And a new function "Delete shaping points" without route calculation in the route menu.

**Comment by 0709wiwiwi** [comment]
Promote in a gpx track one or more existing gpx tkpt. The number of total trkpt and location remains unchanged.

A. Promote a gpx track into a simple gpx navtrack. Change the first and last trackpoint into a via trackpoint without recalculation. New.

Change a single trkpt into a shaping trkpt without recalculation. Fill with shaping trackpoints without recalculation. Exist

Change ALL shaping trkpt into normal trkpt without recalculation. New.

Change a shaping trkpt into a via trkpt without recalculation. Exist.

This allows to (re)spread the shaping trkpt as you wish. Nicely spaced shaping trkpt are more favourable in route (re)calculations.

**Reply by devemux86** [reply]
@0709wiwiwi 

How the (A) is possible without a route calculation? And what is its use?
A normal GPX track cannot be converted to a navigation track (with turn instructions) without route calculation.

All other should be covered by the extra functions I mentioned [above](https://github.com/devemux86/cruiser/discussions/543#discussioncomment-9519124)?

**Reply by devemux86** [reply]
But this is still not a navigation track with turn instructions, it remains a normal GPX track with a list of coordinates,
that may be or not on a routable road. So what is its use?

Via and shaping points can be used on calculated routes.

**Reply by 0709wiwiwi** [reply]
Make track routable as described.

**Reply by devemux86** [reply]
Do you mean a way to convert an imported track to a route with partially calculated sections?

The correct procedure is to import the track as a route with multiple waypoints,
or import it as overlay and construct a route step-by-step by adding waypoints.

**Comment by devemux86** [comment]
Delete one or all shaping points **without route calculation** will be available in the next version:

- "Delete" one shaping point is already available in **waypoint menu** (image 1).

- "Delete shaping points" will be available in **route menu** to delete all shaping points (image 2).
It combines nicely with the existing "Add shaping points" function with adjustable distance.

- These actions are stored in the undo / redo history of the route, so they be can reverted.

- Delete shaping points is done **offline without route calculation**.
If you wish, use the recalculate section or route afterwards.

- Shaping points that can be deleted without route calculation are those that do not define a section with a route profile

<img src="https://github.com/user-attachments/assets/12348727-f19f-4134-a4c0-370d2d494ed2" width="50%"/><img src="https://github.com/user-attachments/assets/55013ebe-41c9-4a66-a291-9f1be215ec6c" width="50%"/>

**Comment by devemux86** [comment]
Delete shaping points without route calculation is available in:
- https://github.com/devemux86/cruiser/discussions/734

See the [link](https://github.com/devemux86/cruiser/discussions/543#discussioncomment-12041010) for instructions.

**Reply by WolfgangZ42** [reply]
Works like a charm! Necessary Shaping points of straight line segments are kept - great! Thanks a lot. 
But if I now want to refill the shaping points, "Add names" is missing ... 😕

![2025-03-13 15_35_35-Workflow_Kurviger_Cruiser_Garmin_en pdf - Foxit PDF Reader](https://github.com/user-attachments/assets/d43a5572-980f-49f1-943d-beec1522dec6). . .![2025-03-13 15_36_06-2025-03-13 15 26 46 jpeg - Windows-Fotoanzeige](https://github.com/user-attachments/assets/8196c917-fa97-420f-90c7-0072e1bdf8c7)

I use these names to synchronize my .kurviger file with the .gpx files from the Garmin users - as described in my workflow: 
[Workflow_Kurviger_Cruiser_Garmin_en.pdf](https://github.com/user-attachments/files/19230929/Workflow_Kurviger_Cruiser_Garmin_en.pdf)

**Reply by devemux86** [reply]
@WolfgangZ42 Shaping points now display the distance / time to start and destination.

The thing is that fixed names with distance label are not safe,
as they are kept in next calculations leading to confused results.

Are they really needed?

**Reply by WolfgangZ42** [reply]
> they are kept in next calculations leading to confused results.

**Good point!** I didn't see that because it is the last step for a "closed" route. "Add names" was an option and could be used with the motto "Trust me, I know what I'm doing". 😎 
Now I'll think about another solution - and possibly come back to you. 😉

**Comment by devemux86** [comment, marked as answer]
- Delete shaping points without route calculation is available in https://github.com/devemux86/cruiser/discussions/757

See the [link](https://github.com/devemux86/cruiser/discussions/543#discussioncomment-12041010) for instructions.

**Comment by MotoUKRider** [comment]
I think I've fallen foul of this new feature on my recent trip.

Wishing to remove several ShapingPoints due to a road-closure, it was more convenient for me to long-press on the map and remove each numbered point (rather than refer to the Waypoints List).  This, didn't recalculate the route - a long-press on the route and invoke a manual recalculation refreshed the route layout.

By the roadside, this was confusing as I didn't know the area, but having just tried a simple local route; Adding a ShapingPoint diverts the route but when I delete it, nothing happens - but if I manually invoke it (long-press on the route line), the re-calculation works.

I have tried "Intelligent" and "Normal" in the "Navigation Settings" but neither have any effect... Perhaps I've found a bug or I've just done something silly.

**Reply by devemux86** [reply]
@MotoUKRider

Shaping point function without calculation:
- Long press on the route and select "Set shaping point"
- Long press on a shaping point and select "Delete"
- Long press on the route and select "Add shaping points"
- Long press on the route and select "Delete shaping points"

Adding waypoints away from the route will cause a route calculation.

**Reply by MotoUKRider** [reply]
@devemux86 That seems a little odd.  If I edit the route in any way, I would like it to be recalculated from my location.
_Adding_ a ShapingPoint, I've confirmed to work - so removing it should also offer the same result?

I was hoping the "Intelligent" option in the Routing menu would solve my issue: To recalculate between _changed_ points on my route (from my current position within it).

Using this old image of mine as an example, if I _planned_ to use the fun road and added (red) ShapingPoints to keep me on it, deleting them manually via a long-press does _not_ alter the route back onto the motorway unless I manually invoke a recalculation.
![routing-points](https://github.com/user-attachments/assets/66b3c849-e9e5-4e6a-81dc-3a0d67aa5317)

In other words, Cruiser is still showing me a route I do not want to take.

**Reply by devemux86** [reply]
@MotoUKRider It always worked that way.

> Adding a ShapingPoint, I've confirmed to work - so removing it should also offer the same result?

What is the same result? Do you add the point on the route or away from the route?
Do you long press **on the route line** or **on the map** near the route line?
The long press menu of the route is different from the long press menu of the map.

All functions that can be done without calculation, will be performed without route calculation.
e.g. set waypoints on the route line, delete waypoints, etc.

At the end simply recalculate the route **or the section**. Otherwise we need to recalculate all the time.
Note: deleting a waypoint does not necessarily mean that the route should not pass through its location.

The functions that require a calculation, like adding a waypoint **away** from the route line,
will trigger a new route calculation.

**Reply by MotoUKRider** [reply]
@devemux86 Adding points is fine, no issue there.  The problem I found was that when I manually delete ShapingPoints via a long-press on the map - I can visually see that the route remains "as planned" but is _not_ the one that I'll be taking, ie: I'm nearly always forcing a slower (fun) route despite using "Fast" BRouter settings.

This works for me 99% of the time, because if for any reason a fun trip needs to be shortened - I can just remove a few points and rely on "get me there fast" routing to take care of things for me!

So, if I my route is: S(tart), 1, 2, 3, 4, 5, E(nd) and I pass ShapingPoint "2", find that the road is closed somewhere along "3".  I remove "3" and would expect an automatic recalculation between Me+4.

**Reply by devemux86** [reply]
@MotoUKRider This is the planned behavior for the "intelligent" routing mode,
because there you can recalculate each section separately.

However, we could change it to calculate the route in the "normal" routing mode,
if there is need by the users.

**Reply by MotoUKRider** [reply]
@devemux86 I've since confirmed that when I make a ShapingPoint deletion, and then press "Navigation" to resume the route, the new changes _are_ reflected (cancelling nav and looking at the overall map in 2D shows this) but making incremental changes while stopped should still produce a change - or at least be an explicit option.

I already have "Intelligent calculation" set in the App Routing options, if that helps.

ie: I delete 1 point, it recalculates and then I can make a decision about the next few points immediately.
It seems that Desktop Cruiser (online BRouter) also does this.

The crux of this issue is that it's _not immediately obvious_ that the route is no longer "optimal" and won't reflect reality unless I invoke navigation and trust it, or manually recalculate.   I hope I'm not being lazy and others appreciate the workflow I'm trying to achieve :laughing:

**Reply by devemux86** [reply]
@MotoUKRider Desktop Cruiser uses online routers, so it uses the "normal" routing mode.

As mentioned above, we can make this change in "normal" routing mode, so it works as expected.
"Intelligent" routing mode better remains as it is now because it provides much greater flexibility.

**Reply by devemux86** [reply]
If want to make it simple, then calculate the route always.

And think again if delete 1 shaping point without calculation is really needed.

**Reply by MotoUKRider** [reply]
I appreciate we may all have different workflows, riding/routing styles and I've certainly encountered a lengthy "number crunch" when loading a particularly "heavy" GPX file into Cruiser (from BaseCamp).  The "intelligent" recalculation (between affected/edited points) seems like the way forward, and certainly the most efficient in terms of CPU, time and user-experience!

Delete vs Remove _does_ sound like a confusing idea, but I get the aim of it.  It just makes sooo much more sense to delete point 23 and have the route recalculate between points 22+24 (rather than the whole lot or nothing).

**Reply by 0709wiwiwi** [reply]
In case of quick changes, please also consider the Cruiser desktop capabilities.

**Reply by MotoUKRider** [reply]
@0709wiwiwi My use of BaseCamp is not the issue here, it's just the tool I have used over the past few years (even before I started with Kv2+Cruiser).  In fact, Kv2+Cruiser have always taken a GPX exported from BaseCamp and loaded it, directly, no conversion, AND with the Shaping+Via-Points mirrored.  Minor routing rules aside, I have no issues with it "navigating" what I have "planned".

But, to address your comment: There's too much missing from Cruiser Desktop to persuade me to change over.  I still use Desktop for theme development and it's an **excellent** map viewer, far easier to work with areas/regions than BaseCamp (even though I compile my own maps).  Lack of simple search (ie: city/town/street) is also a negative.

Without taking this too far off-topic, I think an improved Desktop solution would be an ideal companion - perhaps for those that don't want the online/cloud environment like some other products.  Possibly even a "sync" so that App subscribers can manage data one-click and forego the GPX/JSON import/export routines.  But, as you say, @devemux86 is doing the work solo.  I wish I could get my head around it, but the code environment didn't awaken any creativity in me :cry:  My experience is Visual Basic/Studio and some C++.

**Reply by devemux86** [reply]
We can skip any change here if the recalculate route button becomes available,
as discussed in the other thread:

- https://github.com/devemux86/cruiser/discussions/798#discussioncomment-13170416

**Reply by WolfgangZ42** [reply]
I am confused and only partially understand this discussion! 
Currently everything can be controlled with **setup/routing/advanced/route calculation**, I think. 
Using “_intelligent calculation of route sections_”, then the recalculation is no big deal. 
Adding a shaping point I want a recalculation of the two sections concerned, and that works. 
Deleting a shaping point, I would also expect a recalculation, but this is not the case at the moment. 
I don't need the “new” recalculate route button - because I make all routing changes by long pressing on the route and not via **.../routing/options/route profile**.

**Reply by 0709wiwiwi** [reply]
> Deleting a shaping point, I would also expect a recalculation, but this is not the case at the moment.

@WolfgangZ42.

Roger. Agree.

**Reply by WolfgangZ42** [reply]
![2025-05-16 19 50 49](https://github.com/user-attachments/assets/bbd39bdd-6840-45c9-ab03-9e1e7fe1bc7e)
Works as described - except that in all three cases deleting a shaping point doesn't initiate a recalculation. (As this thread asks for ...)

**Reply by MotoUKRider** [reply]
@WolfgangZ42 Exactly that.  What I [suggested in an earlier post](https://github.com/devemux86/cruiser/discussions/543#discussioncomment-13149877), was the important bit:

_So, if I my route is: S(tart), 1, 2, 3, 4, 5, E(nd) and I pass ShapingPoint "2", find that the road is closed somewhere along "3". I remove "3" and would expect an automatic recalculation between Me+4._

This, quite sensibly, should/would only process the affected "bit" of the route that I've just edited.

**Reply by WolfgangZ42** [reply]
In navigation mode you can use "avoid road block" with an estimated length of blocking. 
Stopping navigation the waypoint "Me" isn't present, so 
1. in "intelligent" mode the remove of "3" should recalculate the segment between "2" (not "Me") and "4". 
2. in "normal" mode there should a recalculation of the whole route occur. 
3. in "manual" mode there should happen nothing, except "recalculate route" is initiated

Unfortunately 1. and 2. don't work this way - as I said, in all three cases no recalculation take place.

**Reply by devemux86** [reply]
@MotoUKRider In navigation everything is handled automatically:
skip next waypoint, select next waypoint to travel, avoid road block.

@WolfgangZ42 Thanks for the summary, this is how it currently works.

The instructions are mentioned [above](https://github.com/devemux86/cruiser/discussions/543#discussioncomment-12041010).

- Note: this thread's idea was to "Delete shaping points" (plural) without calculation.
(route menu)

- During this process, also the delete of 1 shaping point was made without calculation.
(waypoint menu) Useful or not?

**Reply by WolfgangZ42** [reply]
Personally, I would prefer an automatic recalculation to take place after deleting one shaping point in the “intelligent” and “normal” modes. 

Many shaping point changes (deleting, inserting and moving) without recalculation can be carried out with “Route calculation = manual”.

**Reply by devemux86** [reply]
Let's go with what most users expect, delete of a single shaping point is done with calculation.

We'll see if an extra function is really needed and why exactly.
Also "delete" and "remove" translate the same in many languages, which is confusing.
